srml/kton/src/lib.rs

```diff
@@ -236,11 +232,42 @@ impl<T: Trait> Currency<T::AccountId> for Module<T> {
 			return Ok(());
 		}
 
-		let now = <system::Module<T>>::block_number();
-		if locks
-			.into_iter()
-			.all(|l| now >= l.until || new_balance >= l.amount || !l.reasons.intersects(reasons))
-		{
+		// 1. for Lock:
+		//     currently, we only use `WithdrawReasons::all()` in staking module
+		//     so the `!lock.reasons.intersects(reasons)` always be `false`
+		// 2. for Staking Lock:
+		//     make sure the free Kton > the Kton which in Staking
+		// 3. for Unbonding Lock:
+		//     because of the rule 1, only check if is expired
+		let now = <timestamp::Module<T>>::now();
+		let mut can_withdraw_staking = false;
+		let mut can_withdraw_unbonding = false;
+		for lock in locks {
+			match lock.id {
+				Id::Staking(_) => {
+					can_withdraw_staking = !lock.valid_at(&now);
+					if can_withdraw_staking {
+						continue;
+					}
+
+					can_withdraw_staking = new_balance >= lock.amount;
+					if can_withdraw_staking {
+						continue;
+					}
+
+					// due to rule 1
+					can_withdraw_staking = false;
+					// TODO: for complex reasons, in the future
+					// can_withdraw = !lock.reasons.intersects(reasons);
+				}
+				Id::Unbonding(_) => {
+					// due to rule 3
+					can_withdraw_unbonding = !lock.valid_at(&now);
+				}
+			}
+		}
+
+		if can_withdraw_unbonding || can_withdraw_staking {
```

Comment:
It is more clean and decent if this complex logic can be put in one composition Lock, such as Stakking_Unbund_Lock. Then, the balances module does not need to dependent on ID types. and do match checks here.